src/server/mod.rs

```diff
@@ -53,11 +63,230 @@ where
     F: Fn(Request) -> Fut,
     Fut: Future<Output = http_types::Result<Response>>,
 {
+    let mut out_io = io.clone();
+    
     loop {
+        let mut _io = io.clone();
+        // io for headers and normal body reading
+        let (io_primary, mut io_primary_send) = channel_bufread::BufReader::with_capacity(8 * 1024, io.clone());
+        // io used with Body::io
+        let (io_upgrade, mut io_upgrade_send) = channel_bufread::BufReader::with_capacity(8 * 1024, io.clone());
+        
+        let (io_complete_send, io_complete_recv) = oneshot::channel::<()>();
+        
+        use futures_util::sink::SinkExt;
+        async_std::task::spawn(async move {
+            let mut io_end_fut = io_complete_recv.fuse();
+            
+            let (mut body_start, mut n) = {
+                let mut buf: Vec<u8> = Vec::new();
+                buf.resize(8 * 1024, 0);
+                let buf_len = buf.len();
+                let mut pos = 0;
+                // read up to 8kb of headers, the implementation already limits
+                // headers to 8kb
+                let header_end_pos = 'a: loop {
+                    let n = match _io.read(&mut buf[pos..buf_len]).await {
+                        Ok(0) => {
+                            let res = Ok((Arc::new(vec![].into_boxed_slice()), 0usize));
+                            match io_primary_send.send(res).await {
+                                Ok(_) => (),
+                                Err(err) if err.is_disconnected() => (),
+                                Err(err) => unreachable!("Unexpected channel err {}", err),
+                            }
+                            
+                            return;
+                        }
+                        Ok(n) => {
+                            pos += n;
+                            n
+                        }
+                        Err(err) => {
+                            match io_primary_send.send(Err(err)).await {
+                                Ok(_) => (),
+                                Err(err) if err.is_disconnected() => (),
+                                Err(err) => unreachable!("Unexpected channel err {}", err),
+                            }
+                            return
+                        }
+                    };
+                    
+                    if pos < 3 {
+                        continue
+                    }
+                    
+                    for idx in 3..n {
+                        if idx >= 3 && &buf[idx - 3..=idx] == b"\r\n\r\n" {
```

Comment:
Do we need this check?
```suggestion
                        if &buf[idx - 3..=idx] == b"\r\n\r\n" {
```

Reply:
i don't think so